test/src/test-stats.cc

```diff
@@ -0,0 +1,163 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Home for tests focused on some aspect of stats functionality.
+ */
+
+#include <test/support/tdb_catch.h>
+#include "tiledb/api/c_api/context/context_api_internal.h"
+#include <tiledb/common/dynamic_memory/dynamic_memory.h>
+#include <tiledb/common/common.h>
+#include <tiledb/sm/cpp_api/tiledb>
+#include <tiledb/sm/filesystem/vfs.h>
+#include <tiledb/sm/stats/global_stats.h>
+
+#include <string>
+
+using namespace tiledb::sm::stats;
+
+static std::string empty_dumped_stats = { "[\n\n]\n" };
+// Currently there are no stats that are always present.
+// Can, however, envision a time when something (stats calls themselves?) might be
+// accumulated, thus having a forever outstanding stats item, the base
+// version of which might be represented in this base_dumped_stats, rather
+// than the output being totally 'empty'.
+static std::string base_dumped_stats = empty_dumped_stats;
+
+TEST_CASE("Stats inferred registration handling, direct method calls", "[stats]") {
+  // Examine stats output as a reflection of whether allocations might be held beyond a destruction.
+
+  std::string dumped_stats;
+  all_stats.dump(&dumped_stats);
+  CHECK(dumped_stats == base_dumped_stats);
+
+  {
+    std::shared_ptr<Stats> stats { make_shared<Stats>(HERE(), "Test") };
+    
+    all_stats.register_stats(stats);
+
+    all_stats.dump(&dumped_stats);
+    CHECK(dumped_stats == base_dumped_stats);
+
+    stats->create_child("childstats");
+    all_stats.dump(&dumped_stats);
+    CHECK(dumped_stats == base_dumped_stats);
+
+    stats->add_counter("testcounter", 1);
+    all_stats.dump(&dumped_stats);
+    CHECK(dumped_stats != base_dumped_stats);
+  }  
+
+  all_stats.dump(&dumped_stats);
+  CHECK(dumped_stats == base_dumped_stats);
+}
+
+TEST_CASE("Stats registration handling, indirect via Context", "[stats][context]") {
+  // Examine stats output when class Context active as reflection of
+  // whether data allocations might be held beyond a reset and/or registrant destruction.
```

Comment:
I don't like the way this test is written at all. It's relying on external behaviors to exercise a particular execution sequence. External behaviors can change, and we don't want to rely on them. In addition, it's relying on an interaction between `Context` and a global instance of `all_stats`, which has all the problems of testing a global.

It would be far better just to use a local instance as before and directly exercise the call pattern on it that's being tested.